### PR TITLE
Mileston 1 of Mobile v3 swap

### PR DESCRIPTION
**Pull request by YannMarti**
# Grant Application Checklist

- [ ] The [`applications/application-template.md`](../applications/application-template.md) or [`applications/research-template.md`](../applications/research-template.md) has been copied, renamed ( "project_name.md") and updated.
- [ ] An Ethereum address for the payment of the milestones is provided inside the application.  
- [ ] The total funding amount of the project is up to USD $30k for every grant applications.
- [ ] The initial PR contains only one commit (squash if needed before submitting your PR).

### CODE REVIEW

File: milestone-delivery/deliveries/mobile-v3-swap-milestone-1.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# Milestone Delivery :mailbox:
+
+> Only the GitHub account, which is responsible for the pull request of the accepted application is allowed to submit milestones. Don't remove any of the mandatory parts presented in bold letters or as headlines!
+
+**The Ethereum address has been filled out correctly for this milestone and the delivery is according to the official [milestone delivery guidelines](../).**
+
+- **PR Link:** https://github.com/vechain/grant-program/pull/132#issuecomment-1545426640
```
Comment (libotony) [inline milestone-delivery/deliveries/mobile-v3-swap-milestone-1.md:7]:
```suggestion
- **PR Link:** https://github.com/vechain/grant-program/pull/132
```

**Comment by RsTs23** [comment]
Thank you for reviewing our milestone submission and removing the change request label. Is there any specific timeline you are following for the merge of the submission? Our team is eager to continue the work on the project in the frontend but we would like to have our first milestone delivery properly accepted before we start. If there are any further questions or required changes, please let us know.

**Comment by libotony** [comment]
Hi @YannMarti @RsTs23, I apologize for the delay in my response. I am currently working on building your contracts, but I encountered an error. 

```
ParserError: Source "interfaces/IUniswapV3Pool.sol" not found: File not found. Searched the following locations: "".
 --> contracts/HelpFunctions.sol:4:1:
  |
4 | import "../interfaces/IUniswapV3Pool.sol";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


ParserError: Source "interfaces/IERC20.sol" not found: File not found. Searched the following locations: "".
 --> contracts/NFT.sol:6:1:
  |
6 | import "../interfaces/IERC20.sol";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


ParserError: Source "interfaces/IUniswapV3Pool.sol" not found: File not found. Searched the following locations: "".
 --> contracts/NFT.sol:7:1:
  |
7 | import "../interfaces/IUniswapV3Pool.sol";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


ParserError: Source "interfaces/IERC20.sol" not found: File not found. Searched the following locations: "".
 --> contracts/Pool.sol:6:1:
  |
6 | import "../interfaces/IERC20.sol";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


ParserError: Source "interfaces/IUniswapV3FlashCallback.sol" not found: File not found. Searched the following locations: "".
 --> contracts/Pool.sol:7:1:
  |
7 | import "../interfaces/IUniswapV3FlashCallback.sol";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


ParserError: Source "interfaces/IUniswapV3MintCallback.sol" not found: File not found. Searched the following locations: "".
 --> contracts/Pool.sol:8:1:
  |
8 | import "../interfaces/IUniswapV3MintCallback.sol";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


ParserError: Source "interfaces/IUniswapV3Pool.sol" not found: File not found. Searched the following locations: "".
 --> contracts/Pool.sol:9:1:
  |
9 | import "../interfaces/IUniswapV3Pool.sol";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


ParserError: Source "interfaces/IUniswapV3PoolDeployer.sol" not found: File not found. Searched the following locations: "".
  --> contracts/Pool.sol:10:1:
   |
10 | import "../interfaces/IUniswapV3PoolDeployer.sol";
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


ParserError: Source "interfaces/IUniswapV3SwapCallback.sol" not found: File not found. Searched the following locations: "".
  --> contracts/Pool.sol:11:1:
   |
11 | import "../interfaces/IUniswapV3SwapCallback.sol";
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


ParserError: Source "interfaces/IUniswapV3PoolDeployer.sol" not found: File not found. Searched the following locations: "".
 --> contracts/PoolFactory.sol:6:1:
  |
6 | import "../interfaces/IUniswapV3PoolDeployer.sol";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


ParserError: Source "interfaces/IUniswapV3Pool.sol" not found: File not found. Searched the following locations: "".
 --> contracts/Quoter.sol:6:1:
  |
6 | import "../interfaces/IUniswapV3Pool.sol";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


ParserError: Source "interfaces/IERC20.sol" not found: File not found. Searched the following locations: "".
 --> contracts/SwapManager.sol:7:1:
  |
7 | import "../interfaces/IERC20.sol";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


ParserError: Source "interfaces/IUniswapV3Pool.sol" not found: File not found. Searched the following locations: "".
 --> contracts/SwapManager.sol:8:1:
  |
8 | import "../interfaces/IUniswapV3Pool.sol";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


ParserError: Source "interfaces/IUniswapV3Manager.sol" not found: File not found. Searched the following locations: "".
 --> contracts/SwapManager.sol:9:1:
  |
9 | import "../interfaces/IUniswapV3Manager.sol";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^


ParserError: Source "interfaces/IUniswapV3Manager.sol" not found: File not found. Searched the following locations: "".
 --> contracts/new/Utils.sol:6:1:
  |
6 | import "../../interfaces/IUniswapV3Manager.sol";
```

**Comment by YannMarti** [comment]
> Hi @YannMarti @RsTs23, I apologize for the delay in my response. I am currently working on building your contracts, but I encountered an error.
> 
> ```
> ParserError: Source "interfaces/IUniswapV3Pool.sol" not found: File not found. Searched the following locations: "".
>  --> contracts/HelpFunctions.sol:4:1:
>   |
> 4 | import "../interfaces/IUniswapV3Pool.sol";
>   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> 
> ParserError: Source "interfaces/IERC20.sol" not found: File not found. Searched the following locations: "".
>  --> contracts/NFT.sol:6:1:
>   |
> 6 | import "../interfaces/IERC20.sol";
>   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> 
> ParserError: Source "interfaces/IUniswapV3Pool.sol" not found: File not found. Searched the following locations: "".
>  --> contracts/NFT.sol:7:1:
>   |
> 7 | import "../interfaces/IUniswapV3Pool.sol";
>   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> 
> ParserError: Source "interfaces/IERC20.sol" not found: File not found. Searched the following locations: "".
>  --> contracts/Pool.sol:6:1:
>   |
> 6 | import "../interfaces/IERC20.sol";
>   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> 
> ParserError: Source "interfaces/IUniswapV3FlashCallback.sol" not found: File not found. Searched the following locations: "".
>  --> contracts/Pool.sol:7:1:
>   |
> 7 | import "../interfaces/IUniswapV3FlashCallback.sol";
>   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> 
> ParserError: Source "interfaces/IUniswapV3MintCallback.sol" not found: File not found. Searched the following locations: "".
>  --> contracts/Pool.sol:8:1:
>   |
> 8 | import "../interfaces/IUniswapV3MintCallback.sol";
>   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> 
> ParserError: Source "interfaces/IUniswapV3Pool.sol" not found: File not found. Searched the following locations: "".
>  --> contracts/Pool.sol:9:1:
>   |
> 9 | import "../interfaces/IUniswapV3Pool.sol";
>   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> 
> ParserError: Source "interfaces/IUniswapV3PoolDeployer.sol" not found: File not found. Searched the following locations: "".
>   --> contracts/Pool.sol:10:1:
>    |
> 10 | import "../interfaces/IUniswapV3PoolDeployer.sol";
>    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> 
> ParserError: Source "interfaces/IUniswapV3SwapCallback.sol" not found: File not found. Searched the following locations: "".
>   --> contracts/Pool.sol:11:1:
>    |
> 11 | import "../interfaces/IUniswapV3SwapCallback.sol";
>    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> 
> ParserError: Source "interfaces/IUniswapV3PoolDeployer.sol" not found: File not found. Searched the following locations: "".
>  --> contracts/PoolFactory.sol:6:1:
>   |
> 6 | import "../interfaces/IUniswapV3PoolDeployer.sol";
>   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> 
> ParserError: Source "interfaces/IUniswapV3Pool.sol" not found: File not found. Searched the following locations: "".
>  --> contracts/Quoter.sol:6:1:
>   |
> 6 | import "../interfaces/IUniswapV3Pool.sol";
>   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> 
> ParserError: Source "interfaces/IERC20.sol" not found: File not found. Searched the following locations: "".
>  --> contracts/SwapManager.sol:7:1:
>   |
> 7 | import "../interfaces/IERC20.sol";
>   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> 
> ParserError: Source "interfaces/IUniswapV3Pool.sol" not found: File not found. Searched the following locations: "".
>  --> contracts/SwapManager.sol:8:1:
>   |
> 8 | import "../interfaces/IUniswapV3Pool.sol";
>   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> 
> ParserError: Source "interfaces/IUniswapV3Manager.sol" not found: File not found. Searched the following locations: "".
>  --> contracts/SwapManager.sol:9:1:
>   |
> 9 | import "../interfaces/IUniswapV3Manager.sol";
>   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 
> 
> ParserError: Source "interfaces/IUniswapV3Manager.sol" not found: File not found. Searched the following locations: "".
>  --> contracts/new/Utils.sol:6:1:
>   |
> 6 | import "../../interfaces/IUniswapV3Manager.sol";
> ```

Hi @libotony 
That is our mistakes, we accidentally had the interaces in .gitignore. Should be fixed now, sorry about the confusion.

**Comment by libotony** [comment]
> Hi @libotony That is our mistakes, we accidentally had the interfaces in .gitignore. Should be fixed now, sorry about the confusion.

Thanks for the update, the compilation works fine, but I got the following error in running tests:

```
=========================================================================================================================== short test summary info ============================================================================================================================
FAILED tests/Test_SetPos.py::test_inRange - brownie.exceptions.VirtualMachineError: revert: ERC20: insufficient allowance
ERROR tests/Test_updatePos.py::test_updatePositions - brownie.exceptions.VirtualMachineError: revert: ERC20: insufficient allowance
============================================================================================================== 1 failed, 5 passed, 52 warnings, 1 error in 13.39s ==============================================================================================================
```

**Comment by JoaoMorais96** [comment]
Unfortunately, at the time we used Brownie for testing. Although it  was (and still is) a great tool for testing Smart Contracts, it stopped being maintained a while ago. The "brownie.exceptions.VirtualMachineError: revert: ERC20: insufficient allowance" error is well documented but never got solved. It happens sometimes when running multiple testes at once. In reality, Brownie is prone for false failure when doing multiple tests (with different error messages). If you run the 2 failing tests separately, they will pass.

When running all tests using "brownie test":
```
tests\Test_MultiSwap.py .                                                                                      
tests\Test_Quoter.py .                                                                                         
tests\Test_SetPos.py F                                                                                         
tests\Test_SingleSwap.py .                                                                                     
tests\Test_TokensOfOwner.py .                                                                                  
tests\Test_UpdatePos.py .                                                                                      
tests\Test_UserToAllPositions.py F   
============================================== short test summary info ============================================== 
FAILED tests/Test_SetPos.py::test_inRange - brownie.exceptions.VirtualMachineError: revert: ERC20: insufficient all...
FAILED tests/Test_UserToAllPositions.py::test_userToAllPos - brownie.exceptions.VirtualMachineError: revert: typed ...=============================== 2 failed, 5 passed, 52 warnings in 236.35s (0:03:56) ================================
```
When running the 2 failing tests separately:
```
-> brownie test tests/Test_UserToAllPositions.py
tests\Test_UserToAllPositions.py . 
========================================== 1 passed, 8 warnings in 41.12s ===========================================

-> brownie test tests/Test_SetPos.py
tests\Test_SetPos.py . 
========================================== 1 passed, 4 warnings in 22.11s ===========================================
```

**Comment by libotony** [comment]
> Unfortunately, at the time we used Brownie for testing. Although it was (and still is) a great tool for testing Smart Contracts, it stopped being maintained a while ago. The "brownie.exceptions.VirtualMachineError: revert: ERC20: insufficient allowance" error is well documented but never got solved. It happens sometimes when running multiple testes at once. In reality, Brownie is prone for false failure when doing multiple tests (with different error messages). If you run the 2 failing tests separately, they will pass.
> 
> When running all tests using "brownie test":
> 
> ```
> tests\Test_MultiSwap.py .                                                                                      
> tests\Test_Quoter.py .                                                                                         
> tests\Test_SetPos.py F                                                                                         
> tests\Test_SingleSwap.py .                                                                                     
> tests\Test_TokensOfOwner.py .                                                                                  
> tests\Test_UpdatePos.py .                                                                                      
> tests\Test_UserToAllPositions.py F   
> ============================================== short test summary info ============================================== 
> FAILED tests/Test_SetPos.py::test_inRange - brownie.exceptions.VirtualMachineError: revert: ERC20: insufficient all...
> FAILED tests/Test_UserToAllPositions.py::test_userToAllPos - brownie.exceptions.VirtualMachineError: revert: typed ...=============================== 2 failed, 5 passed, 52 warnings in 236.35s (0:03:56) ================================
> ```
> 
> When running the 2 failing tests separately:
> 
> ```
> -> brownie test tests/Test_UserToAllPositions.py
> tests\Test_UserToAllPositions.py . 
> ========================================== 1 passed, 8 warnings in 41.12s ===========================================
> 
> -> brownie test tests/Test_SetPos.py
> tests\Test_SetPos.py . 
> ========================================== 1 passed, 4 warnings in 22.11s ===========================================
> ```

Thank you for the reply, I have run the test separately, and they all get passed. I was wondering if is there a `coverage` for all tests?

**Comment by JoaoMorais96** [comment]
Brownie gave us a very incomplete one (did not recognise most functions used in the testing), so we decided to test all functions and errors that are in the contracts documentations under the folder called "documentation"

**Comment by libotony** [comment]
> Brownie gave us a very incomplete one (did not recognise most functions used in the testing), so we decided to test all functions and errors that are in the contracts documentations under the folder called "documentation". Covers all functions created by us

Thank you for the explaination.

**Review by libotony — APPROVED** [review]

**Comment by NecoSherry** [comment]
Thank you for delivering milestone 1. We will initiate the payment process for milestone 2 promptly. If there are any updates, we will notify you by email. Thank you in advance for your patience and cooperation.

**Pull request merged**